lib/elixir/lib/exception.ex

```diff
@@ -157,6 +157,159 @@ defmodule Exception do
+         {:ok, %{definitions: defs}} <- backend.debug_info(:elixir_v1, module, data, []),
+         {_, kind, _, clauses} <- List.keyfind(defs, {function, arity}, 0) do
+      clauses =
+        for {meta, ex_args, guards, _block} <- clauses do
+          scope = :elixir_erl.definition_scope(meta, kind, function, arity, "nofile")
+          {erl_args, scope} = :elixir_erl_clauses.match(&:elixir_erl_pass.translate_args/2, ex_args, scope)
+          {args, binding} = blame_args(call_args, ex_args, erl_args, [], [])
+          guards = Enum.map(guards, &blame_guard(&1, scope, binding))
+          {args, guards}
+        end
+      {:ok, kind, clauses}
+    else
+      _ -> :error
+    end
+  end
+
+  defp blame_args([call_arg | call_args], [ex_arg | ex_args], [erl_arg | erl_args], binding, ann_args) do
+    {match?, binding} = blame_arg(erl_arg, call_arg, binding)
+    blame_args(call_args, ex_args, erl_args, binding,
+               [blame_wrap(match?, rewrite_arg(ex_arg)) | ann_args])
+  end
+  defp blame_args([], [], [], binding, ann_args) do
+    {Enum.reverse(ann_args), binding}
+  end
+
+  defp blame_arg(erl_arg, call_arg, binding) do
+    binding = List.keystore(binding, :VAR, 0, {:VAR, call_arg})
+    try do
+      {:value, _, binding} = :erl_eval.expr({:match, 0, erl_arg, {:var, 0, :VAR}}, binding, :none)
+      {true, binding}
+    rescue
+      _ -> {false, binding}
+    end
+  end
+
+  defp rewrite_arg(arg) do
+    Macro.prewalk(arg, fn
+      {:%{}, meta, [__struct__: Range, first: first, last: last]} ->
+        {:.., meta, [first, last]}
+      other ->
+        other
+    end)
+  end
+
+  defp blame_guard({{:., _, [:erlang, op]}, meta, [left, right]}, scope, binding)
+       when op == :andalso or op == :orelse do
+    {rewrite_guard_call(op), meta, [
+      blame_guard(left, scope, binding),
+      blame_guard(right, scope, binding)
+    ]}
+  end
+
+  defp blame_guard(ex_guard, scope, binding) do
+    {erl_guard, _} = :elixir_erl_pass.translate(ex_guard, scope)
+    match? =
+      try do
+        {:value, true, _} = :erl_eval.expr(erl_guard, binding, :none)
+        true
+      rescue
+        _ -> false
+      end
+    blame_wrap(match?, rewrite_guard(ex_guard))
+  end
+
+  defp rewrite_guard(guard) do
+    Macro.prewalk(guard, fn
+      {:., _, [:erlang, call]} -> rewrite_guard_call(call)
+      other -> other
+    end)
+  end
+
+  defp rewrite_guard_call(:"orelse"), do: :or
+  defp rewrite_guard_call(:"andalso"), do: :and
+  defp rewrite_guard_call(:"=<"), do: :<=
+  defp rewrite_guard_call(:"/="), do: :!=
+  defp rewrite_guard_call(:"=:="), do: :===
+  defp rewrite_guard_call(:"=/="), do: :!==
+
+  defp rewrite_guard_call(op) when op in [:band, :bor, :bnot, :bsl, :bsr, :bxor],
+    do: {:., [], [Bitwise, op]}
+  defp rewrite_guard_call(op) when op in [:xor, :element, :size],
+    do: {:., [], [:erlang, op]}
+  defp rewrite_guard_call(op),
+    do: op
```

Comment:
Excellent question which I asked myself during the process. If we store the original code, then we need to store the Macro.Env for every clause so we know how to expand it and that would make the :debug_info chunk very expensive in terms of size.